sand/controllers/transform.py

```diff
@@ -0,0 +1,190 @@
+def transform_filter(transform_func, data, tolerance):
+    transformed_data = []
+    exec_tolerance = tolerance
+    for path, value, context in data:
+        tdata = dict()
+        value = value[0]
+        tdata["path"] = path
+        tdata["value"] = value
+        try:
+            result = transform_func(value, context)
+            if type(result) != bool:
+                raise BadRequest("filter transform function must return boolean value")
+            tdata["ok"] = result
+            transformed_data.append(tdata)
+        except Exception as err:
+            (exc, value, tb) = sys.exc_info()
+            tb = tb.tb_next  # Skip *this* frame
+            tdata["error"] = "".join(traceback.format_exception(exc, value, tb))
+            transformed_data.append(tdata)
+            exec_tolerance -= 1
+            if exec_tolerance == 0:
+                return transformed_data
+    return transformed_data
+
+
+def transform_split(transform_func, data, tolerance):
+    transformed_data = []
+    exec_tolerance = tolerance
+    for path, value, context in data:
+        tdata = dict()
+        value = value[0]
+        tdata["path"] = path
+        tdata["value"] = value
+        try:
+            result = transform_func(value, context)
+            if type(result) != list:
+                raise BadRequest("filter transform function must return list")
+            tdata["ok"] = result
+            transformed_data.append(tdata)
+        except Exception as err:
+            (exc, value, tb) = sys.exc_info()
+            tb = tb.tb_next  # Skip *this* frame
+            tdata["error"] = "".join(traceback.format_exception(exc, value, tb))
+            transformed_data.append(tdata)
+            exec_tolerance -= 1
+            if exec_tolerance == 0:
+                return transformed_data
+    return transformed_data
+
+
+def transform_concatenate(transform_func, data, tolerance):
+    transformed_data = []
+    exec_tolerance = tolerance
+    for path, value, context in data:
+        tdata = dict()
+        tdata["path"] = path
+        tdata["value"] = value
+        try:
+            result = transform_func(value, context)
+            tdata["ok"] = result
+            transformed_data.append(tdata)
+        except Exception as err:
+            (exc, value, tb) = sys.exc_info()
+            tb = tb.tb_next  # Skip *this* frame
+            tdata["error"] = "".join(traceback.format_exception(exc, value, tb))
+            transformed_data.append(tdata)
+            exec_tolerance -= 1
+            if exec_tolerance == 0:
+                return transformed_data
+    return transformed_data
+
+
+@transform_bp.route(
+    f"/{transform_bp.name}/<table_id>/transformations", methods=["POST"]
+)
+def transform(table_id: int):
+    table = Table.get_by_id(table_id)
+    table_rows: List[TableRow] = list(
+        TableRow.select().where(TableRow.table == table).order_by(TableRow.index)
+    )
+    transform_type = request.json["type"]
```

Comment:
I think we discussed about validating the inputs in one of the meeting, so we need to do it

Reply:
Completed!